File: feign-api/src/main/java/name/abhijitsarkar/feign/persistence/DefaultIdGenerator.java
/*
 * Copyright (c) 2016, the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * License for more details.
 */

package name.abhijitsarkar.feign.persistence;


import lombok.ToString;
import lombok.extern.slf4j.Slf4j;
import name.abhijitsarkar.feign.Request;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * @author Abhijit Sarkar
 */
@Slf4j
@ToString
@SuppressWarnings({"PMD.ShortMethodName"})
public class DefaultIdGenerator implements IdGenerator {
    @Override
    public String id(Request request) {
        Pattern pattern = Pattern.compile("^(?:/?)([^/]+)(?:.*)$");
        Matcher matcher = pattern.matcher(request.getPath());

        String prefix = matcher.matches() ? matcher.group(1) : "unknown";

        return String.format("%s-%s",
                prefix,
                request.getPath().hashCode());
    }
}

File: feign-api/src/main/scala/name/abhijitsarkar/feign/persistence/DefaultIdGenerator.scala
/*
 * Copyright (c) 2016, the original author or authors.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; version 3 of the License.
 *  *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * A copy of the GNU General Public License accompanies this software,
 * and is also available at http://www.gnu.org/licenses.
 *
 */

package name.abhijitsarkar.feign.persistence

import name.abhijitsarkar.feign.Request

/**
  * @author Abhijit Sarkar
  */
class DefaultIdGenerator extends IdGenerator {
  val pattern = """^(?:/?)([^/]+)(?:.*)$""".r

  override def id(request: Request): String = {
    val prefix = pattern.findFirstMatchIn("/abc/xyz")
      .map(_ group 1)
      .getOrElse("unknown")

    s"${prefix}-${request.path.hashCode}"
  }
}
